dev_tools/oneapi/convert_source.sh

```diff
@@ -0,0 +1,429 @@
+echo "${GINKGO_LICENSE_BEACON}*/" >> "${OUTPUT_FILE}"
+rm "${GLOBAL_FILE}"
+rm "${EMBED_FILE}"
+while IFS='' read -r line; do
+    if [[ "${line}" =~ $DEVICE_CODE_SYNTAX ]]; then
+        # hold the command to easy replace
+        echo "// ${line}" >> ${EMBED_FILE}
+        device_file="${BASH_REMATCH[1]}"
+        [ "${EXTRACT_KERNEL}" == "true" ] && echo "/**** ${device_file} - start ****/" >> "${EMBED_FILE}"
+        cat "${ROOT_DIR}/${device_file}" >> "${EMBED_FILE}"
+        [ "${EXTRACT_KERNEL}" == "true" ] && echo "/**** ${device_file} - end ****/" >> "${EMBED_FILE}"
+        if [ -n "${DEVICE_FILE}" ]; then
+            DEVICE_FILE="${DEVICE_FILE};"
+        fi
+        DEVICE_FILE="${DEVICE_FILE}${device_file}"
+    else
+        echo "${line}" >> "${EMBED_FILE}"
+    fi
+done < "${UNFORMAT_FILE}"
+${CLANG_FORMAT} -style=file "${EMBED_FILE}" > "${FORMAT_FILE}"
+${SCRIPT_DIR}/add_host_function.sh "${FORMAT_FILE}" > "${EMBED_HOST_FILE}"
+
+CONFIG_REGEX="constexpr *Config *([a-zA-Z0-9_]*) *= *([a-zA-Z][a-zA-Z_:,\(\) ]*)"
+while IFS='' read -r line; do
+    if [[ "${IN_SYNTAX}" = "false" ]] && [[ "${line}" =~ ${CONFIG_REGEX} ]]; then
+        config_name="${BASH_REMATCH[1]}"
+        config_content="${BASH_REMATCH[2]}"
+        echo "${line}" >> "${OUTPUT_FILE}"
+        echo "constexpr auto ${config_name}_list = ::gko::syn::value_list<Config, ${config_name}>();" >> "${OUTPUT_FILE}"
+    elif [[ "${line}" =~ ${KERNEL_SYNTAX_START} ]] || [[ "${IN_SYNTAX}" = "true" ]]; then
+        temp="${temp} ${line}"
+        IN_SYNTAX="true"
+        if [[ "${line}" =~ ${FUNCTION_END} ]]; then
+            IN_SYNTAX="false"
+            modified=""
+            if [[ "${temp}" = *"/*KEEP*/"* ]]; then
+                modified="${temp/\/\*KEEP\*\//}"
+            else
+                # change <<<>>> to (grid, block, dynamic, queue)
+                modified=$(convert_syntax "$temp")
+            fi
+            # echo "temp -> convert"
+            # echo "$temp"
+            # echo "$modified"
+            echo "${modified}" >> "${OUTPUT_FILE}"
+            temp=""
+        fi
+    else
+        echo "${line}" >> "${OUTPUT_FILE}"
+    fi
+done < "${EMBED_HOST_FILE}"
+
+# Other fix on OUTPUT_FILE
+# dim3 -> dim3_t (for easy replace)
+# this_thread_block -> this_thread_block_t
+# tiled_partition -> tiled_partition_t
+# thread_id.cuh -> use local
+# cooperative_group.cuh -> use local
+replace_regex="s/dim3/dim3_t/g"
+replace_regex="${replace_regex};s/this_thread_block/this_thread_block_t/g"
+replace_regex="${replace_regex};s/this_grid/this_grid_t/g"
+replace_regex="${replace_regex};s/tiled_partition/tiled_partition_t/g"
+replace_regex="${replace_regex};s/thread::/thread_t::/g"
+replace_regex="${replace_regex};s/bitonic_sort/bitonic_sort_t/g"
+replace_regex="${replace_regex};s/reduction_array/reduction_array_t/g"
+replace_regex="${replace_regex};s/cuda\/components\/thread_ids.cuh/trick\/thread_ids.hpp/g"
```

Comment:
```replace_regex="${replace_regex};s:cuda/components/thread_ids.cuh:trick/thread_ids.hpp:g"```

Reply:
it needs to escape `/`, right?

Reply:
I still need to escape the `.`, right?
although it may not change the behavior in general